Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/LocationUpdates.hs

```diff
@@ -115,3 +121,28 @@ whenWithLocationUpdatesLock driverId f = do
       throwError (HitsLimitError 5)
   where
     lockKey = "DriverLocationUpdate:DriverId-" <> driverId.getId
+
+performSafetyCheck :: Id Person -> [LatLong] -> Flow ()
+performSafetyCheck driverId batchWaypoints = do
+  ride <- runInReplica $ QRide.getActiveByDriverId driverId >>= fromMaybeM (RideForDriverNotFound $ getId driverId)
+  transporterConfig <- MTC.findByMerchantOpCityId ride.merchantOperatingCityId >>= fromMaybeM (TransporterConfigNotFound ride.merchantOperatingCityId.getId)
+  now <- getLocalCurrentTime transporterConfig.timeDiffFromUtc
+  let midnightTime = UTCTime (utctDay now) 0
+      nextMidNight = addUTCTime nominalDay midnightTime
+      startTime = addUTCTime (secondsToNominalDiffTime transporterConfig.nightSafetyStartTime) midnightTime
+      endTime = addUTCTime (secondsToNominalDiffTime transporterConfig.nightSafetyEndTime) midnightTime
+      key = searchRequestKey (getId ride.id)
+  when (now >= midnightTime && now <= endTime || now >= startTime && now <= nextMidNight) $ do
+    routeInfo :: Maybe RI.RouteInfo <- Redis.get key
+    case routeInfo >>= (.points) of
+      Just estimatedRoute -> do
+        when (not ride.safetyAlertTriggered && (checkForDeviation transporterConfig.safetyRouteDeviationThreshold estimatedRoute batchWaypoints 0 1)) $ do
```

Comment:
This check should be at top level with timing checks.

Reply:
Why deviationCount is 1, it should be atleast 3 to avoid outliers. Sometimes single point location update become outlier.